### PR TITLE
Use an executor for the processing of the agents

### PR DESCRIPTION
**Pull request by Shredder121**
I wanted to wait with this one last, but I think it's also good to get some comments/feedback in.
So here goes.

This makes the point where agents are started unified, and cleans up the `//noinspection InfiniteLoopStatement while (true) {`, each one start a Thread, (correctly) configuring it, etc. mantra.

Caveats/notes:

- [x] This makes the starting/handling/rate-scheduling/stopping universal.
- [x] `log.info("starting xxx")` should be either 1) removed, or 2) be accompanied by `log.info("finished xxx")` or 3) be changed in logging level.
- [x] Decide on whether to use `scheduleWithFixedDelay` (which was more akin to the previous code, since there were always fixed sleep intervals) or `scheduleAtFixedRate` (which makes more sense (?) when dealing with agents/cleanup units of work).
- [x] Deal with potential `Throwable`s killing off the agents.
- [ ] Decide whether a single worker thread is enough, or if a larger threadpool or a cached threadpool is better, so 2, 3, 4 threads? Or let the cached threadpool handle it?
- [ ] Decide whether we use `AGENTS.awaitTermination()` in the shutdown, or do we just let it shutdown while we do other shutdown-y things?
- [ ] Needs a lot of eyes on the code. I have been running this code for about/over a week or two, but that's different from when it would go live for thousands of servers.
- [ ] Do not merge until all checks have been ticked, since other issues can be raised 😨. (I don't want to be *that guy who broke everything*)

### CODE REVIEW

File: FredBoat/src/main/java/fredboat/agent/FredBoatAgent.java [hunk 1]
```diff
@@ -32,7 +32,7 @@
 public abstract class FredBoatAgent implements Runnable {
 
     private static final ScheduledExecutorService AGENTS = Executors.newSingleThreadScheduledExecutor(runnable -> {
-        Thread thread = new Thread(runnable);
+        Thread thread = new Thread(runnable, "agent runner thread");
```
Comment (Shredder121) [inline FredBoat/src/main/java/fredboat/agent/FredBoatAgent.java]: Stub, if anyone knows of a *nicer* name?

**Comment by Frederikam** [comment]
I'd say `scheduleAtFixedRate` is fine as our agents all take a long while between executions. The time they run is still much smaller than in the production environment.

File: FredBoat/src/main/java/fredboat/agent/FredBoatAgent.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public abstract class FredBoatAgent implements Runnable {
+
+    private static final ScheduledExecutorService AGENTS = Executors.newSingleThreadScheduledExecutor(runnable -> {
+        Thread thread = new Thread(runnable, "agent runner thread");
+        thread.setDaemon(true);
```
Comment (Shredder121) [inline FredBoat/src/main/java/fredboat/agent/FredBoatAgent.java]: Also, one could argue since we now properly shutdown the agents, that we do not need the worker(s) to be daemon threads.

Comment (Frederikam) [inline FredBoat/src/main/java/fredboat/agent/FredBoatAgent.java]: 👍

File: FredBoat/src/main/java/fredboat/agent/FredBoatAgent.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+package fredboat.agent;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public abstract class FredBoatAgent implements Runnable {
+
+    private static final ScheduledExecutorService AGENTS = Executors.newSingleThreadScheduledExecutor(runnable -> {
+        Thread thread = new Thread(runnable, "agent runner thread");
```
Comment (Shredder121) [inline FredBoat/src/main/java/fredboat/agent/FredBoatAgent.java]: In fact **not** outdated, *GitHub*, still a stub, if someone knows a better name.

Or a naming scheme, if we go for multiple threads in the thread pool.

**Comment by napstr** [comment]
So this uses one Thread to do everything? We've had the situation that a thread got stuck doing his task. What happens here in that situation?

**Comment by Frederikam** [comment]
>So this uses one Thread to do everything? We've had the situation that a thread got stuck doing his task. What happens here in that situation?

We'll probably want to use a pool of threads. The production bot already has several thousands of threads running. A few more won't hurt.

**Comment by napstr** [comment]
Actually in that situation we had to go over a thread dump and find the threads by name. So I think it's important to keep the agent's tasks independent from each other, and have a distinct name.

**Comment by Frederikam** [comment]
Can we not just change the name when a task starts? We can just add it to the constructor of the abstract agent.

**Comment by Shredder121** [comment]
>A few more won't hurt.

That.. is a way of thinking about it.

> We've had the situation that a thread got stuck doing his task. 

> So I think it's important to keep the agent's tasks independent from each other, and have a distinct name.

Yes distinct threads is a good idea.

> Can we not just change the name when a task starts? We can just add it to the constructor of the abstract agent.

Yeah, we can do that, I'm now changing the code, and testing it again while listening to the music it produces.

**Comment by Shredder121** [comment]
> Actually in that situation we had to go over a thread dump and find the threads by name. So I think it's important to keep the agent's tasks independent from each other, and have a distinct name.

We can also introduce a ThreadGroup, so the spawned threads are grouped?
But maybe later, now just correctness.

**Comment by Shredder121** [comment]
Thread group is maybe not necessary, but at least the name now reflects the agent that's running.

**Comment by Frederikam** [comment]
ea3e280082429d36644e9baf146372ce19c7607b causes a conflict with this PR

**Comment by Shredder121** [comment]
I'll rebase later, now need to do something else.

But thanks for letting me know 😄

**Comment by napstr** [comment]
This is important:

http://code.nomad-labs.com/2011/12/09/mother-fk-the-scheduledexecutorservice/

TL;DR: If scheduled tasks encounter an exception, they will be cancelled _silently_.

**Comment by Shredder121** [comment]
I know, that's why I already kept the try-catches.
The only place I didn't is the CarbonitexAgent, because the only play it could throw was already try-catched.
Or do you want to have it anyway?

**Comment by napstr** [comment]
Idk. Things failing silently by default really rustles my jimmies
```
try {}
catch (Throwable t) {}
```
¯\\\_(ツ)\_/¯

**Comment by Shredder121** [comment]
`Throwable`? You mad man!
(But I can do that :/)

**Comment by knight-ryu12** [comment]
ffs use Exception e, It is good for you

**Comment by Frederikam** [comment]
Why silence the exceptions? It's a pain to debug if we don't have a trace.

**Comment by Shredder121** [comment]
How is this?

**Comment by Shredder121** [comment]
The individual agents already have try-catches, so this is just a catch-all for if something goes _horribly_ wrong.

**Comment by napstr** [comment]
Hey idk if this is overkill, but a command to check the status of all FredBoat agents, and to cancel and start them, wouldn't that be neat to have for us admins?

**Comment by napstr** [comment]
Not necessarily a command, I would take colored indicators on some kind of admin dashboard if I could, but that's not a thing.

**Comment by Shredder121** [comment]
Ohh yeah I can whip something up

**Comment by napstr** [comment]
About those uncaught exceptions, TIL there is this thing:

https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.UncaughtExceptionHandler.html

**Comment by Shredder121** [comment]
Yeah, that's useful if they run in a `Thread`, but they run on a thread pool, and I'm pretty sure the worker threads gobble the exception, and don't report it up.
It would be pretty expensive to let the threads actually die after each exception and then create a new one.
<details>
<summary>Snippet taken from `FutureTask.runAndReset()`</summary>

```java
try {
    Callable<V> c = callable;
    if (c != null && s == NEW) {
        try {
            c.call(); // don't set result
            ran = true;
        } catch (Throwable ex) {
            setException(ex);
        }
    }
} finally {
```

</details>

**Comment by Shredder121** [comment]
So, status:
Currently set to use 2 threads, should not have any runaway tasks in the agents(the 'problem size' is pretty small -- playing players, active (potentially dead) shards)

The tasks won't ever die, because we catch all throwables (and at least log it, so we don't shoot ourselves in the foot by silently ignoring any problem)

> Decide whether we use AGENTS.awaitTermination() in the shutdown, or do we just let it shutdown while we do other shutdown-y things?

Do we just assume the pool will have been closed? do we block and wait or not?

P.S.
Did we kill the shard watchdog agent yet? should I pull that?

**Comment by napstr** [comment]
About the shard watchdog: Let's circumvent having to make an ultimate decision and make it a feature flag :upside_down_face: 

We can close the pool in the shutdown hook. Even if it's only running daemons etc that should still be a best practice.

**Comment by Shredder121** [comment]
> We can close the pool in the shutdown hook. Even if it's only running daemons etc that should still be a best practice.

Well it's already closing in the shutdown hook, but do we block and wait until it's finished, or just assume it will go away and do other shutdown-y things in the meantime?
It's doing the latter now, but I want opinions.

Anyway, amount of threads is fine? (1 for the frequent things, the other for backup and having some parallelism?)

**Comment by napstr** [comment]
Put it on a timeout of 30sec or something imo, just in case it gets stuck.
Amount of threads looks fine to me.

**Comment by Shredder121** [comment]
So I rebased this one, but noticed there is not a way anymore to reach the ShardWatchdogAgent.

I see we put it under a feature flag, is that guaranteed safe then?
I assume so, but would like a sign off that it's now good, since we don't have a way to reach it.

**Comment by napstr** [comment]
Looks good to me. I'm just not sure I like the log spam. Probably better to move these to the debug level?
image

**Comment by Shredder121** [comment]
Debug level it is then, one moment.

**Comment by napstr** [comment]
Tbh even the debug level is annoying (which I'm using locally quite often).

Is trace level viable? Or any other idea?

**Comment by Shredder121** [comment]
Trace is doable, sure
Does it end up in any log file?
If not just remove it

**Comment by napstr** [comment]
Nope, we don't log traces at all. This doesn't feel like a useful log anyways since we track the last execution time and it's walled up against exceptions properly too.

**Comment by Shredder121** [comment]
Then I vote for removal of starting/running, but keep the thread name adjustment for if the exceptions happen and something else is logged
